### DISCUSSION TITLE
How to mock a module loaded using `require()`

### DISCUSSION THREAD

**Discussion by cefn** [vitest-dev/vitest#3134] [Q&A] [answered]
I realised (eventually) that vi.mock silently omits mocking modules which are loaded using `require()`. 

> Note to self - maybe an unused vi.mock should cause vitest to fail a suite?

Is there a standard answer for mocking CommonJS style imports that is compatible with vitest? I'm assuming that blending `jest` tooling with `vitest` is a bad idea so jest.mock isn't a feasible option.

Everything else about my test suite (in a commonJS repository) is working well with no transpilation or build step - I am seamlessly using Typescript to author tests against functions defined using commonjs formalisms. However, the one occasion where I need to mock a required module is turning out to be very costly. To fulfil the requirements of `vi.mock` I have to introduce `import` and `export` keywords in the target files, migrate the repo `package.json` to `"type":"module"` and so on leading to a cascade of changes for a single mock. In this particular context, lots of other tooling assumes commonjs making this approach impossible.

What options do I have for mocking a module loaded with `require()` when using vitest. I really don't want to go back to Jest for this one feature.

**Comment by sheremet-va** [comment]
Vitest doesn't intercept `require`, so it is not possible to mock a module that was loaded with `require`

**Reply by cefn** [reply]
Thanks, Vladimir. I'm aware that the `vi.mock` support in Vitest doesn't intercept require. 

My question is whether there is any other mechanism, library or approach that's compatible with Vitest (not part of Vitest) and that intercepts require. 

Perhaps there's a theoretical reason why this is impossible because of how Vitest runs, which I would be interested to hear. This would at least clarify what I need to do (switch back to using jest for anything authored in CommonJS that needs module mocking for tests).

**Reply by cefn** [reply]
By way of an example, there is https://www.npmjs.com/package/mock-require but that hasn't been updated in 4 years, perhaps because CommonJS is an outdated standard? Hence wondering if there's a standard choice of strategy that people adopt that avoids any obvious pitfalls. Perhaps that older module is still fine as it doesn't need updates.

**Reply by djMax** [reply]
"outdated standard" made me LOL. It is no more outdated than "Javascript" is outdated.

**Reply by jiovanmichel** [reply]
Use vi.spyOn

_BeforeEach / BeforeAll_
vi.spyOn(class, 'method').mockReturnValue('mock value');

_AfterEach / AfterAll_
vi.restoreAllMocks();

**Comment by cefn** [comment, marked as answer]
Adopting `mock-require` in the test suite was an effective workaround.

**Reply by elringus** [reply]
No need to pull a package for this; here's minimal sample to mock require in node:

```js
function mock(mockedUri, stub) {
    Module._load_original = Module._load;
    Module._load = function (uri, parent) {
        if (uri === mockedUri) return stub;
        return Module._load_original(uri, parent);
    };
}
```

**Reply by rjumatov** [reply]
This works pretty well but there is a flaw. If the require import statements are at the top this custom mock function won't intercept.
I had to make inline require imports to successfully mock those packages.

**Reply by sheremet-va** [reply]
If you put it inside `vi.hoisted`, then it will be executed before your imports

**Reply by rjumatov** [reply]
Thank you four your quick reply. I get a ReferenceError since ```Module from 'module'``` needs to be imported before `vi.hoisted`.
Is there a way to execute selected imports before `vi.hoisted`?

**Reply by sheremet-va** [reply]
You can import modules inside vi.hoisted. Don't forget to await it

**Reply by rjumatov** [reply]
Thanks, works with dynamic imports and `vi.hoisted`. Very neat! :grimacing:

**Reply by binodov** [reply]
What exactly was the solution? How do we lever the above function in a vitest test?

**Reply by tim-trewartha** [reply]
Thanks for this helpful discussion - I found `mock-require` does the job but I'd like to avoid it if possible. I'm not too sure how to achieve the same with `vi.hoisted`. My issue is that a node-module imported in a shared (transpiled) package is imported using `require`, and thus I am unable to intercept using Vitest.

**Comment by haayhappen** [comment]
Can somebody post some code snipped with a working mock-require implementation? That would be very helpful 👍

**Reply by rjumatov** [reply]
I have this scenario to mock:
```js
const {ClassToMock} = require('@toBeMocked');
```

And am mocking it successfully like that at the top of my test class:
```js
vi.hoisted(
    () =>
        void mock('@toBeMocked', {
            ClassToMock: class ClassToMock {
                getSomething = () => yourReturnValue;
            },
        })
);

async function mock(mockedUri, stub) {
    const {Module} = await import('module');

    Module._load_original = Module._load;
    Module._load = (uri, parent) => {
        if (uri === mockedUri) return stub;
        return Module._load_original(uri, parent);
    };
}
```

If you have exported functions to mock your stub would look different but other than that this works pretty well.

**Reply by viceice** [reply]
simpler:
```ts
vi.hoisted(()=> { 
  require.cache.fs = { exports: fsMock } as never;
});

**Comment by jmussman** [comment]
In case anybody comes late to the party...

@sm3sher provided the answer but it will only work with a single CommonJS module. I needed more. I created a full-blown package that shims the module loader and allows mocking of multiple CommonJS modules. At GitHub: https://github.com/jmussman/vitest-mock-commonjs, and at npmjs.com: https://www.npmjs.com/package/vitest-mock-commonjs.

More importantly the README delves deep into how it all works if you want to know, and there is an open-source stand-alone example that depends on the package at https://github.com/jmussman/auth0-block-idp-signup.

**Reply by deninpaul** [reply]
You are the GOAT

**Reply by jmussman** [reply]
Thank you, I am glad it helped!

**Comment by rlindner81** [comment]
Since [mock-require](https://www.npmjs.com/package/mock-require) is the accepted answer here:
That module was out of maintenance already at the time it was accepted here. The associated [repo](https://github.com/boblauer/mock-require) has been archived on Mar 15, 2021.

There are plenty of alternative implementations and the underlying code is very small in scope with about 100 lines of code. So it's feasible to rig up your own mock system in this style.
